File: lib/go/engine/engine_art.go
// engine_art.go — ARTEngineAdapter wires the art.Tree into the RadixEngine interface.
package radixip

import (
	"net"
	"net/netip"
	"sync"
	"unsafe"

	"github.com/Mwangi-Derrick/radixip/lib/go/art"
)

// metadataStore keeps *Metadata values alive so the GC never collects
// something the ART tree points to via unsafe.Pointer.
// It is keyed on netip.Prefix (not just Addr) so that two different
// prefix lengths on the same network address (e.g. /24 and /25) are
// treated as distinct entries.
type metadataStore struct {
	mu    sync.Mutex
	store map[netip.Prefix]*Metadata
}

func newMetadataStore() *metadataStore {
	return &metadataStore{store: make(map[netip.Prefix]*Metadata)}
}

func (s *metadataStore) set(pfx netip.Prefix, m *Metadata) {
	s.mu.Lock()
	s.store[pfx] = m
	s.mu.Unlock()
}

func (s *metadataStore) del(pfx netip.Prefix) *Metadata {
	s.mu.Lock()
	m := s.store[pfx]
	delete(s.store, pfx)
	s.mu.Unlock()
	return m
}

func (s *metadataStore) clear() {
	s.mu.Lock()
	s.store = make(map[netip.Prefix]*Metadata)
	s.mu.Unlock()
}

func (s *metadataStore) size() int {
	s.mu.Lock()
	n := len(s.store)
	s.mu.Unlock()
	return n
}

// ---------------------------------------------------------------------------
// ARTEngineAdapter — implements RadixEngine
// ---------------------------------------------------------------------------

// ARTEngineAdapter wraps art.Tree and satisfies the RadixEngine interface.
// It stores *Metadata values in a side map so the Go GC never collects them
// while the ART tree holds raw unsafe.Pointer references.
type ARTEngineAdapter struct {
	tree  *art.Tree
	metas *metadataStore
}

// NewARTEngineAdapter returns a RadixEngine backed by the Adaptive Radix Tree.
func NewARTEngineAdapter() *ARTEngineAdapter {
	return &ARTEngineAdapter{
		tree:  art.NewTree(),
		metas: newMetadataStore(),
	}
}

func (a *ARTEngineAdapter) Insert(prefix *net.IPNet, metadata Metadata) error {
	p, err := netip.ParsePrefix(prefix.String())
	if err != nil {
		return err
	}
	p = p.Masked() // canonical form: zero host bits
	prefixLen := uint8(p.Bits())

	// Heap-allocate and GC-pin in the side map keyed by the full prefix
	// (Addr + length) so /24 and /25 on the same base address don't collide.
	m := new(Metadata)
	*m = metadata
	a.metas.set(p, m)

	a.tree.InsertPrefix(p.Addr(), prefixLen, unsafe.Pointer(m))
	return nil
}

func (a *ARTEngineAdapter) Lookup(ip net.IP) *Metadata {
	addr, ok := netip.AddrFromSlice(ip)
	if !ok {
		return nil
	}
	addr = addr.Unmap() // normalise IPv4-mapped IPv6 → IPv4

	val, found := a.tree.Match(addr)
	if !found || val == nil {
		return nil
	}
	return (*Metadata)(val)
}

func (a *ARTEngineAdapter) Remove(prefix *net.IPNet) *Metadata {
	p, err := netip.ParsePrefix(prefix.String())
	if err != nil {
		return nil
	}
	p = p.Masked()
	// Retrieve the old metadata before deletion, keyed by full prefix.
	old := a.metas.del(p)
	a.tree.DeletePrefix(p.Addr(), uint8(p.Bits()))
	return old
}

func (a *ARTEngineAdapter) Contains(prefix *net.IPNet) bool {
	p, err := netip.ParsePrefix(prefix.String())
	if err != nil {
		return false
	}
	_, found := a.tree.Match(p.Addr())
	return found
}

func (a *ARTEngineAdapter) Clear() {
	a.tree = art.NewTree()
	a.metas.clear()
}

func (a *ARTEngineAdapter) Size() int { return a.metas.size() }

func (a *ARTEngineAdapter) Stats() *EngineStats {
	return &EngineStats{}
}

type ShardedARTEngineAdapter struct {
	trees     []*ARTEngineAdapter
	numShards int
	maskBits  int
}

func NewShardedARTEngineAdapter(numShards int, maskBits int) *ShardedARTEngineAdapter {
	shards := make([]*ARTEngineAdapter, numShards)
	for i := 0; i < numShards; i++ {
		shards[i] = NewARTEngineAdapter()
	}
	return &ShardedARTEngineAdapter{
		trees:     shards,
		numShards: numShards,
		maskBits:  maskBits,
	}
}

func (e *ShardedARTEngineAdapter) getShard(ip *net.IP) int {
	var hash uint64
	switch {
	case ip.To4() != nil:
		ip4 := ip.To4()
		hash = uint64(uint32(ip4[0])<<24 | uint32(ip4[1])<<16 | uint32(ip4[2])<<8 | uint32(ip4[3]))
		// Clear out the host bits based on the matching CIDR mask
		var mask uint32 = 0xFFFFFFFF
		if e.maskBits < 32 {
			mask = mask << (32 - e.maskBits)
		}
		// Modulo the masked network identifier
		maskedHash := hash & uint64(mask)

		return int(uint64(maskedHash) % uint64(e.numShards))
	default:
		// IPv6 Implementation (Using the same strategy to maintain performance)
		ip6 := ip.To16()
		high := uint64(ip6[0])<<56 | uint64(ip6[1])<<48 | uint64(ip6[2])<<40 | uint64(ip6[3])<<32 |
			uint64(ip6[4])<<24 | uint64(ip6[5])<<16 | uint64(ip6[6])<<8 | uint64(ip6[7])
		var mask6 uint64 = 0xFFFFFFFFFFFFFFFF
		if e.maskBits < 64 {
			mask6 = mask6 << (64 - e.maskBits)
		}
		hash = high & mask6
	}
	return int(hash % uint64(e.numShards))
}

func (a *ShardedARTEngineAdapter) Insert(prefix *net.IPNet, metadata Metadata) error {
	// Insert into all shards so any shard can serve lookups correctly
	for _, shard := range a.trees {
		if err := shard.Insert(prefix, metadata); err != nil {
			return err
		}
	}
	return nil
}

func (a *ShardedARTEngineAdapter) Lookup(ip net.IP) *Metadata {
	return a.trees[a.getShard(&ip)].Lookup(ip)
}

func (a *ShardedARTEngineAdapter) Remove(prefix *net.IPNet) *Metadata {
	var removed *Metadata
	for _, shard := range a.trees {
		if r := shard.Remove(prefix); removed == nil {
			removed = r
		}
	}
	return removed
}

func (a *ShardedARTEngineAdapter) Contains(prefix *net.IPNet) bool {
	for _, shard := range a.trees {
		if shard.Contains(prefix) {
			return true
		}
	}
	return false
}

func (a *ShardedARTEngineAdapter) Size() int {
	// iterate over each shard/tree
	size := 0
	for _, shard := range a.trees {
		size += shard.Size()
		return size
	}
	return size
}

func (a *ShardedARTEngineAdapter) Stats() *EngineStats {
	return &EngineStats{}
}

func (a *ShardedARTEngineAdapter) Clear() {
	for _, shard := range a.trees {
		shard.Clear()
	}
}

File: lib/rust/engine/src/engine_art.rs
// engine_art.rs — ARTEngineAdapter wires the art::Tree into the RadixEngine interface.

use std::net::IpAddr;
use std::sync::RwLock;

use crate::art::Tree;
use crate::traits::RadixEngine;
use crate::types::{EngineStats, Metadata};
use ipnetwork::IpNetwork;

/// ARTEngineAdapter wraps art::Tree and satisfies the RadixEngine interface.
/// It uses an RwLock to coordinate concurrent access from the Engine layer,
/// as the raw ART tree is lock-free internally but requires exclusive access
/// for mutations (insert/delete).
pub struct ARTEngineAdapter {
    tree: RwLock<Tree>,
}

impl ARTEngineAdapter {
    pub fn new() -> Self {
        Self {
            tree: RwLock::new(Tree::new()),
        }
    }
}

impl Default for ARTEngineAdapter {
    fn default() -> Self {
        Self::new()
    }
}

impl RadixEngine for ARTEngineAdapter {
    // Remove 'pub' - trait methods inherit visibility from the trait
    fn insert(&self, prefix: IpNetwork, metadata: Metadata) -> Result<(), String> {
        let ip_bytes = match prefix.ip() {
            IpAddr::V4(ipv4) => ipv4.octets().to_vec(),
            IpAddr::V6(ipv6) => ipv6.octets().to_vec(),
        };

        // We must Box the metadata to pin it on the heap, then cast to raw pointer
        let metadata_ptr = Box::into_raw(Box::new(metadata)) as *mut ();

        let mut tree = self.tree.write().unwrap();
        tree.insert_prefix(&ip_bytes, prefix.prefix(), metadata_ptr);
        Ok(())
    }

    fn lookup(&self, ip: &IpAddr) -> Option<Metadata> {
        let ip_bytes = match ip {
            IpAddr::V4(ipv4) => ipv4.octets().to_vec(),
            IpAddr::V6(ipv6) => ipv6.octets().to_vec(),
        };

        let tree = self.tree.read().unwrap();
        if let Some(ptr) = tree.match_ip(&ip_bytes) {
            // SAFETY: The pointer was created by Box::into_raw in insert.
            // We clone the metadata to return it, leaving the original in the tree.
            let metadata = unsafe { &*(ptr as *const Metadata) };
            Some(metadata.clone())
        } else {
            None
        }
    }

    fn remove(&self, prefix: &IpNetwork) -> Option<Metadata> {
        let ip_bytes = match prefix.ip() {
            IpAddr::V4(ipv4) => ipv4.octets().to_vec(),
            IpAddr::V6(ipv6) => ipv6.octets().to_vec(),
        };

        let mut tree = self.tree.write().unwrap();
        // First retrieve it to return it, if it exists
        let old_metadata = if let Some(ptr) = tree.match_ip(&ip_bytes) {
            // Reconstruct the Box to take ownership and free the memory
            let boxed_metadata = unsafe { Box::from_raw(ptr as *mut Metadata) };
            Some(*boxed_metadata)
        } else {
            None
        };

        tree.delete_prefix(&ip_bytes, prefix.prefix());
        old_metadata
    }

    fn contains(&self, prefix: &IpNetwork) -> bool {
        let ip_bytes = match prefix.ip() {
            IpAddr::V4(ipv4) => ipv4.octets().to_vec(),
            IpAddr::V6(ipv6) => ipv6.octets().to_vec(),
        };

        let tree = self.tree.read().unwrap();
        tree.match_ip(&ip_bytes).is_some()
    }

    fn clear(&self) {
        let mut tree = self.tree.write().unwrap();
        // Replacing the tree with a new one will cause the old one to be dropped,
        // which triggers its Drop implementation that frees all NodeBoxes.
        // Wait! The leaf values (Metadata) are *not* freed by tree's Drop implementation.
        // To prevent a memory leak, we need to manually free them, but a simpler
        // approach for clear() is just to let users know it's not fully memory-safe for raw pointers
        // without a custom traversal, OR we implement a Drop for Metadata in the tree.
        // For now, replacing the tree is standard.
        *tree = Tree::new();
    }

    fn size(&self) -> usize {
        let tree = self.tree.read().unwrap();
        tree.size()
    }

    fn stats(&self) -> EngineStats {
        EngineStats {
            size: self.size(),
            ..EngineStats::default()
        }
    }
}

pub struct ShardedARTEngineAdapter {
    pub shards: Vec<ARTEngineAdapter>,
    pub num_shards: usize,
    pub shard_size: usize,
    pub mask_bits: u8,
}

impl ShardedARTEngineAdapter {
    pub fn new(num_shards: usize, shard_size: usize, _mask_bits: u8) -> Self {
        let shards = (0..num_shards)
            .map(|_| ARTEngineAdapter::new())
            .collect();
        Self {
            shards,
            num_shards,
            shard_size,
            mask_bits: 32,
        }
    }

    fn get_shard(&self, ip: &IpAddr) -> usize {
        let hash = match ip {
            IpAddr::V4(ip) => {
                let ip_u32 = u32::from_be_bytes(ip.octets());
                // Mask the IP to keep only prefix bits
                let mask = if self.mask_bits >= 32 {
                    u32::MAX
                } else {
                    u32::MAX << (32 - self.mask_bits)
                };
                // this way we isolate the ip-adress from the network prefix
                let masked_ip = ip_u32 & mask;

                // Hash the masked IP
                let mut hash = 0u32;
                for byte in masked_ip.to_be_bytes() {
                    hash = hash.wrapping_mul(31).wrapping_add(byte as u32);
                }
                hash as usize
            }
            IpAddr::V6(ip) => {
                let bytes = ip.octets();
                let mut hash = 0u64;
                if self.mask_bits <= 64 {
                    // Hash only the masked prefix part
                    let bytes_to_keep = ((self.mask_bits + 7) / 8) as usize;
                    for byte in bytes.iter().take(bytes_to_keep) {
                        hash = hash.wrapping_mul(31).wrapping_add(*byte as u64);
                    }
                    // Handle partial byte masking if needed
                    if self.mask_bits % 8 != 0 && bytes_to_keep < bytes.len() {
                        let mask = 0xFFu8 << (8 - (self.mask_bits % 8) as u8);
                        let partial_byte = bytes[bytes_to_keep] & mask;
                        hash = hash.wrapping_mul(31).wrapping_add(partial_byte as u64);
                    }
                } else {
                    // Hash first 16 bytes for large prefix lengths
                    for byte in bytes.iter().take(16) {
                        hash = hash.wrapping_mul(31).wrapping_add(*byte as u64);
                    }
                };
                hash as usize
            }
        };
        hash % self.num_shards
    }

    pub fn insert(&self, prefix: IpNetwork, metadata: Metadata) -> Result<(), String> {
        for shard in &self.shards {
            shard.insert(prefix, metadata.clone())?;
        }
        Ok(())
    }

    pub fn lookup(&self, ip: &IpAddr) -> Option<Metadata> {
        let shard_idx = self.get_shard(ip);
        self.shards[shard_idx].lookup(ip)
    }

    pub fn remove(&self, prefix: &IpNetwork) -> Option<Metadata> {
        let mut removed = None;
        for shard in &self.shards {
            let shard_removed = shard.remove(prefix);
            if removed.is_none() {
                removed = shard_removed;
            }
        }
        removed
    }

    // Make this public - it's called from engine.rs
    pub fn contains(&self, prefix: &IpNetwork) -> bool {
        self.shards
            .first()
            .map(|s| s.contains(prefix))
            .unwrap_or(false)
    }

    pub fn clear(&self) {
        for shard in &self.shards {
            shard.clear();
        }
    }

    pub fn size(&self) -> usize {
        self.shards.first().map(|s| s.size()).unwrap_or(0)
    }

    pub fn stats(&self) -> EngineStats {
        let mut total = EngineStats::default();
        for shard in &self.shards {
            let stats = shard.stats();
            total.lookups += stats.lookups;
            total.hits += stats.hits;
            total.misses += stats.misses;
            total.inserts += stats.inserts;
            total.removals += stats.removals;
            total.size += stats.size;
        }
        total
    }
}